### ISSUE TITLE
How to check Schema before make_storage?

### ISSUE THREAD

**Issue by Zahir-Khan** [fnc12/sqlite_orm#727]
I have an evolving database that is created on an external application. I use this library to read it. How do I make sure not to crash the library on make_storage if some columns are missing? Sorry for the noob question I'm a c# guy.

I was going to do as simple as try catch code but it will be better to check first.

**Comment by fnc12** [comment]
you won't crash anything inside `make_storage` call cause `make_storage` creates an object in C++ just like `std::make_tuple` or `std::make_unique`. `make_storage` does not open SQLite database. Calling `sync_schema` can alter database's table. Just skip calling it.

**Comment by Zahir-Khan** [comment]
@fnc12 ah yes. Let me rephrase. How do I check if the schema matches between the library and the database without altering the database or crashing the application because I will be querying non existent columns (not calling sync schema because I just want to read the database).

Right now I have a column to check the minimum database version required before proceeding to read it but this is not a solid approach for me.

**Comment by fnc12** [comment]
call `storage.sync_schema_simulate()`. It will return result the same `sync_schema` does but will not alter any table

**Comment by fnc12** [comment]
@Zahir-Khan is the issue actual?

**Comment by Zahir-Khan** [comment]
Thank you. This works. But there is another related problem. My primary key column has NOT NULL constraint. If I run the `sync_schema_simulate()` I am getting result map as `sync_schema_result::dropped_and_recreated` instead of sync_schema_result::already_in_sync everything is matching except for `NOT NULL` constraint. How do I define it in my make_column defenition?

I looked at the code but wasn't able to find any way to do this.

**Comment by fnc12** [comment]
what field is mapped to this PK? Any default C++ class field has `NOT NULL` in `sqlite_orm` except nullable types like `std::unique_ptr` and `std::optional`

**Comment by Zahir-Khan** [comment]
```
inline auto EcProjectReader::InitStorage(const std::string &path)
{
	return make_storage(path,
		make_table("EclMetadataModel",
			make_column("DatabaseVersion", &EclMetadataModel::DatabaseVersion),
			make_column("MinimumCompatibilityVersion", &EclMetadataModel::MinimumCompatibilityVersion),
			make_column("ID", &EclMetadataModel::ID, primary_key(), autoincrement())),
```

It is ID which is just an int. The table created by the external application has NOT NULL constraint for this column.

This is what it shows in DB Browser.

```
CREATE TABLE "EclMetadataModel" (
"DatabaseVersion" integer ,
"MinimumCompatibilityVersion" integer ,
"ID" integer primary key autoincrement not null )
```

**Comment by fnc12** [comment]
@Zahir-Khan let me check it soon. I'll write you back

**Comment by fnc12** [comment]
I've checked it. Table recreation happens cause 'DatabaseVersion' and 'MinimumCompatibilityVersion' has not `NOT NULL` constraint in your SQL query:
```sql
CREATE TABLE "EclMetadataModel" (
"DatabaseVersion" integer ,
"MinimumCompatibilityVersion" integer ,
"ID" integer primary key autoincrement not null )
```
You need either add `NOT NULL` to these columns or change your type for `EclMetadataModel::DatabaseVersion` and `EclMetadataModel::MinimumCompatibilityVersion` from `int` to `std::optional<int>`

**Comment by Zahir-Khan** [comment]
Yup, how do I add Not null in marke_column ? I wasn't able to find a constraint in the library for it.

**Comment by Zahir-Khan** [comment]
```
make_column expects at least two arguments: a column's name and it's mapped member pointer. For example:

make_column("name", &UserType::name);
You don't have to specify type explicitly: it is deduced from member's type you specify as the second parameter. So if you need to create column first_name TEXT NOT NULL just pass make_column("first_name", &User::firstName); where std::string firstName; is a member of User class with std::string type. If you need to create column type_id INTEGER NOT NULL pass make_column("type_id", &User::typeId); where int typeId; is member of User class.

If you need to add some constraints to column like PRIMARY KEY, AUTOINCREMENT or DEFAULTyou can use primary_key, autoincrement or default_value functions respectively. Examples:

id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL - make_column("id", &UserType::id, autoincrement(), primary_key()) comment TEXT DEFAULT 'user' NOT NULL - make_column("comment", &UserType::comment, default_value("user"))

NULL and NOT NULL is deduced from member type. std::string, int and any other type except std::shared_ptr and std::unique_ptr are mapped as NOT NULL and std::shared_ptr and std::unique_ptr are mapped as NULL. Of course you can set your own type as nullable if you wish. For example boost::optional. To do so you have to create a specialization for type_is_nullable class. Example code of using type_is_nullable can be found in Examples/nullable_enum_binding.cpp.
```

This is from the wiki
https://github.com/fnc12/sqlite_orm/wiki/Making-a-storage

`sync_schema_simulate()` should be able to infer the `NOT NULL` constraint automatically then...

I have no clue about C++ or I would have helped out. Looks like there is a need for a seperate constraint like `not_null()` similar to `primarykey()` etc.

**Comment by fnc12** [comment]
You don't need to add NOT NULL to make_column call explicitly cause it has not null already by default implicitly. All types except std::unique_ptr, std::shared_ptr and std::optional automatically treated as not null types inside any storage. What you need is to modify your database and add NOT NULL to the first two columns cause they are nullable right now.

**Comment by fnc12** [comment]
You can use `std::unique_ptr` instead of `std::optional` with C++11/14 easy. You have a choice. It is a very simple issue

**Comment by Zahir-Khan** [comment]
Just tried it. No Bueno. I have 80+ columns to deal with and most of the reading code has already been already written. 
Cant spend too much time on this anymore. I can easily incorporate try catch for the first table with the database version info which would rarely change. Just need simple check if i should proceed to use the database or ask the user to update their application.

Thank you for your help.
